src/scheduler/node_info.c

```diff
@@ -160,6 +160,63 @@
 /* name of the last node a job ran on - used in smp_dist = round robin */
 static char last_node_name[PBS_MAXSVRJOBID];
 
+void
+query_node_info_chunk(th_data_query_ninfo *data)
+{
+	struct batch_status *nodes;
+	struct batch_status *cur_node;
+	node_info **ninfo_arr;
+	server_info *sinfo;
+	node_info *ninfo;
+	int i;
+	int nidx;
+	int start;
+	int end;
+	int num_nodes_chunk;
+
+	nodes = data->nodes;
+	sinfo = data->sinfo;
+	start = data->sidx;
+	end = data->eidx;
+	num_nodes_chunk = end - start + 1;
+
+	if ((ninfo_arr = (node_info **) malloc((num_nodes_chunk + 1) * sizeof(node_info *))) == NULL) {
+		log_err(errno, __func__, MEM_ERR_MSG);
```

Comment:
It will be great if you can clearly demarcate which of the functions are meant to be called from the worker threads and which are the main thread only. That way future developers can keep them safe. It might be beneficial to move all "threaded" functions to one area of the code file (say the end of the file) - with a clear demarcation in between the threaded and non-threaded code. (even better if you can move it to a separate file)

If this function is threaded, then the log_err might not be safe. We must used a thread local buffer for the log message